tests/bwc/test_recovery.py

```diff
@@ -0,0 +1,491 @@
+        rolling upgrade.
+        """
+        cluster = self._new_cluster(path.from_version, nodes)
+        cluster.start()
+        with connect(cluster.node().http_url, error_trace=True) as conn:
+            c = conn.cursor()
+            c.execute('''
+                      create table doc.test(id int primary key, data text) clustered into 1 shards with(
+                       "unassigned.node_left.delayed_timeout" = '100ms', "number_of_replicas" = 2)
+                      ''')
+
+            inserts = [(i, str(random.randint)) for i in range(0, 100)]
+            c.executemany('''insert into doc.test(id, data) values (?, ?)''', inserts)
+
+            c.execute('refresh table doc.test')
+
+            # upgrade to mixed cluster
+            self._upgrade_cluster(cluster, path.to_version, random.randint(1, nodes - 1))
+
+            if random.choice([True, False]):
+                time.sleep(5)
+                self._assert_is_green(conn, 'doc', 'test')
+
+            # update the data in a mixed cluster
+            updates = [(i, str(random.randint)) for i in range(0, 100)]
+
+            res = c.executemany(
+                'insert into doc.test(id, data) values(?, ?) on conflict(id) do update set data = excluded.data',
+                updates)
+            self.assertEqual(len(res), 100)
+            for result in res:
+                self.assertEqual(result['rowcount'], 1)
+
+            if random.choice([True, False]):
+                c.execute('refresh table doc.test')
+
+            # upgrade fully to the new version
+            self._upgrade_cluster(cluster, path.to_version, nodes)
+
+            updates = [(i, str(random.randint)) for i in range(0, 100)]
+            res = c.executemany(
+                'insert into doc.test(id, data) values(?, ?) on conflict(id) do update set data = excluded.data',
+                updates)
+            self.assertEqual(len(res), 100)
+            for result in res:
+                self.assertEqual(result['rowcount'], 1)
+
+    def _test_operation_based_recovery(self, path, nodes):
+        """
+        Tests that we should perform an operation-based recovery if there were
+        some but not too many uncommitted documents (i.e., less than 10% of
+        committed documents or the extra translog) before we upgrade each node.
+        This is important when we move from the translog based to retention leases
+        based peer recoveries.
+        """
+
+        cluster = self._new_cluster(path.from_version, nodes)
+        cluster.start()
+
+        with connect(cluster.node().http_url, error_trace=True) as conn:
+            c = conn.cursor()
+            c.execute('''
+                        create table doc.test(x int) clustered into 1 shards with( "number_of_replicas" = 2,
+                        "soft_deletes.enabled" = true)
+                        ''')
+
+            time.sleep(3)
+            self._assert_is_green(conn, 'doc', 'test')
```

Comment:
Maybe you could build the `assert_busy` for this? 
A fixed sleep is  likely to cause flaky test failures on less powerful jenkins slaves

See the `wait_for_active_shards` https://github.com/crate/crate-qa/blob/632331fed7911996709fbef37d5fbdeb11b15128/src/crate/qa/tests.py#L84-L116

or the `wait_until` and `Timeout`:

https://github.com/mfussenegger/cr8/blob/24e6ba4d30a70edd29f64e1a55829e8d5d22d0eb/cr8/run_crate.py#L132-L166